tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/BsonSerialization/IBsonWriterProxy.cs

```diff
@@ -0,0 +1,225 @@
+﻿// <copyright file="IBsonWriterProxy.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.MongoDb.BsonSerialization;
+
+/// <summary>
+/// Duck Typing interface proxy for: https://github.com/mongodb/mongo-csharp-driver/blob/v2.8.x/src/MongoDB.Bson/IO/IBsonWriter.cs
+/// </summary>
+internal interface IBsonWriterProxy : IDuckType
```

Comment:
Wouldn't we have less to implement if we inherited from BsonWriter instead of IBsonWriter?

Reply:
I don't quite follow 🤔 We need to implement `IBsonWriter` using reverse duck typing, so we can pass it to the `serializer`. I don't _think_ we can reverse duck type anything other than an `interface`, though I confess, I haven't tried 🤔 

For reference, we currently have something like this (not sure if this makes sense)

```mermaid
classDiagram
    IBsonWriter <|-- ReverseDuckType_IBsonWriterProxy : Implements
    IBsonWriterProxy <|-- DuckType_IBsonWriterProxy  : Implements
    MongoBsonWriter
    class ReverseDuckType_IBsonWriterProxy{
      +MongoBsonWriter instance
    }
    class MongoBsonWriter{
      +DuckType_IBsonWriterProxy proxy
    }
    class DuckType_IBsonWriterProxy  {
      +JsonWriter instance
    }
```

i.e 
- we duck type the `JsonWriter` as `IBsonWriterProxy`
- we pass the proxy to a new instance of `MongoBsonWriter`
- we reverse duck type `MongoBsonWriter` as `JsonWriter`

Reply:
So yeah - looks like I was wrong. That _would_ have made it far simpler, you're right 👍  Unfortunately we need to pass parameters to the constructor, and I'm not sure exactly how to do that with duck typing. Looking through the tests, I don't think we can, so I don't think we can reverse duck type `JsonWriter` or `BsonWriter` unfortunately.